File: mxit-compressor.js
import { pathToFileURL } from "node:url";

const ABBREVIATIONS = Object.freeze([
  { source: "thank you", target: "ty" },
  { source: "going", target: "goin" },
  { source: "because", target: "coz" },
  { source: "tonight", target: "2nyt" },
  { source: "tomorrow", target: "2moro" },
  { source: "today", target: "2day" },
  { source: "before", target: "b4" },
  { source: "later", target: "l8r" },
  { source: "please", target: "plz" },
  { source: "thanks", target: "thx" },
  { source: "sorry", target: "sry" },
  { source: "okay", target: "k" },
  { source: "great", target: "gr8" },
  { source: "where", target: "whr" },
  { source: "when", target: "wen" },
  { source: "what", target: "wt" },
  { source: "why", target: "y" },
  { source: "how", target: "hw" },
  { source: "your", target: "ur" },
  { source: "they", target: "dey" },
  { source: "them", target: "dem" },
  { source: "we", target: "we" },
  { source: "know", target: "knw" },
  { source: "think", target: "thnk" },
  { source: "want", target: "wnt" },
  { source: "have", target: "hv" },
  { source: "make", target: "mk" },
  { source: "build", target: "bld" },
  { source: "create", target: "crt" },
  { source: "need", target: "nd" },
  { source: "system", target: "sys" },
  { source: "memory", target: "mem" },
  { source: "model", target: "mdl" },
  { source: "local", target: "lcl" },
  { source: "message", target: "msg" },
  { source: "information", target: "info" },
  { source: "with", target: "w/" },
  { source: "this", target: "dis" },
  { source: "that", target: "dat" },
  { source: "and", target: "n" },
  { source: "you", target: "u" },
  { source: "are", target: "r" },
  { source: "too", target: "2also" },
  { source: "ate", target: "8" },
  { source: "for", target: "4" },
  { source: "to", target: "2" },
  { source: "the", target: "da" },
  { source: "am", target: "m" }
]);

const SORTED_ABBREVIATIONS = [...ABBREVIATIONS].sort((left, right) => {
  return (
    right.source.length - left.source.length ||
    right.target.length - left.target.length
  );
});

const REVERSE_ABBREVIATIONS = [...new Map(
  SORTED_ABBREVIATIONS.map(({ source, target }) => [target, source])
).entries()]
  .map(([target, source]) => ({ source: target, target: source }))
  .sort((left, right) => {
    return (
      right.source.length - left.source.length ||
      right.target.length - left.target.length
    );
  });

function escapeRegex(value) {
  return value.replace(/[.*+?^${}()|[\]\\]/g, "\\$&");
}

function replaceWholePhrase(input, source, target) {
  const pattern = new RegExp(
    `(?<![a-z0-9])${escapeRegex(source)}(?![a-z0-9])`,
    "g"
  );
  return input.replace(pattern, target);
}

function tidySpacing(input) {
  return input
    .replace(/\s+([?!.,;:])/g, "$1")
    .replace(/([([{])\s+/g, "$1")
    .replace(/\s+([)\]}])/g, "$1")
    .replace(/[ \t]+/g, " ")
    .trim();
}

function applyAbbreviations(input, table) {
  return table.reduce((result, entry) => {
    return replaceWholePhrase(result, entry.source, entry.target);
  }, input);
}

export function mxitCompress(rawText = "") {
  const original = String(rawText ?? "");
  const working = tidySpacing(original.toLowerCase());
  const text = applyAbbreviations(working, SORTED_ABBREVIATIONS);
  const originalLength = working.length;
  const compressedLength = text.length;
  const ratio = originalLength === 0
    ? 1
    : Number((compressedLength / originalLength).toFixed(3));

  return {
    text,
    ratio,
    originalLength,
    compressedLength
  };
}

export function mxitDecompress(compressedText = "") {
  const working = tidySpacing(String(compressedText ?? "").toLowerCase());
  return applyAbbreviations(working, REVERSE_ABBREVIATIONS);
}

export { ABBREVIATIONS };

const isDirectRun = Boolean(process.argv[1]) &&
  pathToFileURL(process.argv[1]).href === import.meta.url;

if (isDirectRun) {
  const sample = "I am going to build the system tomorrow because I need to know how it works";
  const compressed = mxitCompress(sample);

  console.assert(
    compressed.text === "i m goin 2 bld da sys 2moro coz i nd 2 knw hw it works",
    `Unexpected compressed output: ${compressed.text}`
  );
  console.assert(
    compressed.compressedLength < compressed.originalLength,
    "Compression should shorten the sample"
  );

  const decompressed = mxitDecompress(compressed.text);
  console.assert(
    decompressed.includes("tomorrow") && decompressed.includes("system"),
    `Unexpected decompressed output: ${decompressed}`
  );

  console.log("mxit-compressor.js tests passed");
}

File: mobile/lib/mxit_compressor.dart
class CompressionResult {
  const CompressionResult({
    required this.text,
    required this.ratio,
    required this.originalLength,
    required this.compressedLength,
  });

  final String text;
  final double ratio;
  final int originalLength;
  final int compressedLength;

  Map<String, dynamic> toJson() {
    return {
      'text': text,
      'ratio': ratio,
      'originalLength': originalLength,
      'compressedLength': compressedLength,
    };
  }
}

final List<MapEntry<String, String>> abbreviations = List.unmodifiable([
  const MapEntry('thank you', 'ty'),
  const MapEntry('going', 'goin'),
  const MapEntry('because', 'coz'),
  const MapEntry('tonight', '2nyt'),
  const MapEntry('tomorrow', '2moro'),
  const MapEntry('today', '2day'),
  const MapEntry('before', 'b4'),
  const MapEntry('later', 'l8r'),
  const MapEntry('please', 'plz'),
  const MapEntry('thanks', 'thx'),
  const MapEntry('sorry', 'sry'),
  const MapEntry('okay', 'k'),
  const MapEntry('great', 'gr8'),
  const MapEntry('where', 'whr'),
  const MapEntry('when', 'wen'),
  const MapEntry('what', 'wt'),
  const MapEntry('why', 'y'),
  const MapEntry('how', 'hw'),
  const MapEntry('your', 'ur'),
  const MapEntry('they', 'dey'),
  const MapEntry('them', 'dem'),
  const MapEntry('we', 'we'),
  const MapEntry('know', 'knw'),
  const MapEntry('think', 'thnk'),
  const MapEntry('want', 'wnt'),
  const MapEntry('have', 'hv'),
  const MapEntry('make', 'mk'),
  const MapEntry('build', 'bld'),
  const MapEntry('create', 'crt'),
  const MapEntry('need', 'nd'),
  const MapEntry('system', 'sys'),
  const MapEntry('memory', 'mem'),
  const MapEntry('model', 'mdl'),
  const MapEntry('local', 'lcl'),
  const MapEntry('message', 'msg'),
  const MapEntry('information', 'info'),
  const MapEntry('with', 'w/'),
  const MapEntry('this', 'dis'),
  const MapEntry('that', 'dat'),
  const MapEntry('and', 'n'),
  const MapEntry('you', 'u'),
  const MapEntry('are', 'r'),
  const MapEntry('too', '2also'),
  const MapEntry('ate', '8'),
  const MapEntry('for', '4'),
  const MapEntry('to', '2'),
  const MapEntry('the', 'da'),
  const MapEntry('am', 'm'),
]);

final List<MapEntry<String, String>> _sortedAbbreviations = [...abbreviations]
  ..sort((left, right) {
    final bySource = right.key.length.compareTo(left.key.length);
    if (bySource != 0) {
      return bySource;
    }
    return right.value.length.compareTo(left.value.length);
  });

final List<MapEntry<String, String>> _reverseAbbreviations = (() {
  final seenTargets = <String>{};
  final reversed = <MapEntry<String, String>>[];

  for (final entry in _sortedAbbreviations) {
    if (seenTargets.add(entry.value)) {
      reversed.add(MapEntry(entry.value, entry.key));
    }
  }

  reversed.sort((left, right) {
    final bySource = right.key.length.compareTo(left.key.length);
    if (bySource != 0) {
      return bySource;
    }
    return right.value.length.compareTo(left.value.length);
  });

  return List.unmodifiable(reversed);
})();

String _replaceWholePhrase(String input, String source, String target) {
  final pattern = RegExp(
    '(^|[^a-z0-9])${RegExp.escape(source)}(?=[^a-z0-9]|$)',
  );

  return input.replaceAllMapped(pattern, (match) {
    final prefix = match.group(1) ?? '';
    return '$prefix$target';
  });
}

String _tidySpacing(String input) {
  return input
      .replaceAllMapped(RegExp(r'\s+([?!.,;:])'), (match) => match.group(1)!)
      .replaceAllMapped(RegExp(r'([([{])\s+'), (match) => match.group(1)!)
      .replaceAllMapped(RegExp(r'\s+([)\]}])'), (match) => match.group(1)!)
      .replaceAll(RegExp(r'[ \t]+'), ' ')
      .trim();
}

String _applyAbbreviations(
  String input,
  List<MapEntry<String, String>> table,
) {
  var result = input;
  for (final entry in table) {
    result = _replaceWholePhrase(result, entry.key, entry.value);
  }
  return result;
}

CompressionResult mxitCompress(String rawText) {
  final working = _tidySpacing(rawText.toLowerCase());
  final text = _applyAbbreviations(working, _sortedAbbreviations);
  final originalLength = working.length;
  final compressedLength = text.length;
  final ratio = originalLength == 0
      ? 1.0
      : double.parse((compressedLength / originalLength).toStringAsFixed(3));

  return CompressionResult(
    text: text,
    ratio: ratio,
    originalLength: originalLength,
    compressedLength: compressedLength,
  );
}

String mxitDecompress(String compressedText) {
  final working = _tidySpacing(compressedText.toLowerCase());
  return _applyAbbreviations(working, _reverseAbbreviations);
}
